File: Trees/Tree.py
class Tree:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

    def insert(self, data):
        new = Tree(data)
        if data < self.data:
            if not self.left:
                self.left = new
                print('%d added to the left of %d' %
                      (new.data, self.data))
                return
            self.left.insert(data)
        else:
            if not self.right:
                self.right = new
                print('%d added to the right of %d' %
                      (new.data, self.data))
                return
            self.right.insert(data)

    def search(self, key):
        if key < self.data:
            if not self.left:
                print('\n%d not found' % key)
                return
            self.left.search(key)
        elif key > self.data:
            if not self.right:
                print('\n%d not found' % key)
                return
            self.right.search(key)
        else:
            print('\n%d found' % key)

    def PrintTree(self):
        if self.left:
            self.left.PrintTree()
        print(self.data),
        if self.right:
            self.right.PrintTree()

Tree1 = Tree(30)
Tree1.insert(20)
Tree1.insert(25)
Tree1.insert(15)
Tree1.insert(40)
Tree1.insert(35)
Tree1.PrintTree()
Tree1.search(20)
Tree1.search(10)
Tree1.search(200)

File: geeksforgeeks/DSTutorials/Trees/Tree.c
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>

struct Node *root = NULL;
struct Node *curr = NULL;
struct Node *searched_node = NULL;

struct Node
{
  int data;
  struct Node *left;
  struct Node *right;
};

struct Node *addNode(int data, struct Node *curr)
{
  struct Node *new = (struct Node *)malloc(sizeof(struct Node));
  new->data = data;
  new->left = new->right = NULL;
  if (!curr)
  {
    // printf("Added %d\n",data);
    curr = new;
    return new;
  }
  if (data < curr->data)
    curr->left = addNode(data, curr->left);
  else
    curr->right = addNode(data, curr->right);
}

void inorder(struct Node *curr)
{
  if (curr)
  {
    inorder(curr->left);
    printf("%d ", curr->data);
    inorder(curr->right);
  }
  else
    return;
}

void preorder(struct Node *curr)
{
  if (curr)
  {
    printf("%d \n", curr->data);
    preorder(curr->left);
    preorder(curr->right);
  }
  else
    return;
}
struct Node *findRightMost(struct Node *node)
{
  if (node->right)
  {
    return findRightMost(node->right);
  }
  return node;
}

void findNode(struct Node *node, int key, struct Node **found)
{
  // printf("\nAt:%d", node->data);
  if (!node)
  {
    return;
  }
  findNode(node->left, key, found);
  if (node->data == key)
    *found = node;
  findNode(node->right, key, found);
}

void delete (struct Node *node, int key)
{
  struct Node *last = findRightMost(node);
  struct Node *to_delete = NULL;
  findNode(node, key, &to_delete);
  to_delete->data = last->data;
  free(last);
}

int main()
{
  root = addNode(50, root);
  addNode(30, root);
  addNode(20, root);
  addNode(40, root);
  addNode(70, root);
  addNode(60, root);
  addNode(80, root);
  addNode(81, root);
  addNode(10, root);
  addNode(25, root);
  addNode(82, root);
  printf("Inorder traversal\n");
  inorder(root);
  // printf("Preorder traversal\n");
  // preorder(root);
  // struct Node *temp = findRightMost(root);
  // printf("\nRightmost:%d\n", temp->data);
  findNode(root, 25, &searched_node);
  if (searched_node)
    printf("\nFound data:%d", searched_node->data);
  delete (root, 40);
  printf("\nInorder traversal\n");
  inorder(root);
  printf("\n");
}
